reports/glides_validator.livemd

```diff
@@ -145,10 +201,426 @@ table_ui =
+          # Same ID as in the second part of key
+          # e.g. "G-10351"
+          id: Common.vehicle_id(),
+          vehicle: %{
+            optional(:stop_id) => Commmon.stop_id(),
+            # unused for this report
+            optional(:multi_carriage_details) => [
+              %{
+                # index of this carriage in the sequence, starting with 1
+                carriage_sequence: pos_integer,
+                # carriage ID maybe?
+                label: String.t(),
+                occupancy_status: occupancy_status,
+                orientation: carriage_orientation
+              }
+            ],
+            current_status: status,
+            # index of the stop it's at or approaching
+            current_stop_sequence: integer,
+            # unused for this report
+            position: %{
+              # compass direction in degrees
+              bearing: integer,
+              latitude: float,
+              longitude: float,
+              speed: float
+            },
+            timestamp: Commmon.timestamp(),
+            trip: %{
+              direction_id: 0 | 1,
+              last_trip: boolean,
+              revenue: boolean,
+              route_id: Commmon.route_id(),
+              schedule_relationship: String.t(),
+              start_date: Commmon.date_string(),
+              start_time: Commmon.time_string(),
+              trip_id: Commmon.trip_id()
+            },
+            vehicle: %{
+              id: Common.vehicle_id(),
+              # Formed by joining multi_carriage_details[].label with "-"
+              # e.g. "3682-3849"
+              label: String.t()
+            }
+          }
+        }
+
+  # "IN_TRANSIT_TO" | "STOPPED_AT" | "INCOMING_AT"
+  @type status :: String.t()
+
+  # "NO_DATA_AVAILABLE" | "STANDING_ROOM_ONLY" | "FEW_SEATS_AVAILABLE"
+  # | "MANY_SEATS_AVAILABLE" | "CRUSHED_STANDING_ROOM_ONLY"
+  @type occupancy_status :: String.t()
+
+  # "AB" | "BA"
+  @type carriage_orientation :: String.t()
+end
+```
+
+## Data mashing
+
+### First, some utility functions:
+
+```elixir
+# Streams all values from an ETS table. (Assuming table's objects are {key, value} 2-tuples)
+stream_values = fn table ->
+  :ets.first(table)
+  |> Stream.iterate(fn key -> :ets.next(table, key) end)
+  |> Stream.take_while(fn key -> key != :"$end_of_table" end)
+  |> Stream.map(fn key -> :ets.lookup_element(table, key, 2) end)
+end
+
+on_line? =
+  case line do
+    "Green Line" -> fn route_id -> String.starts_with?(route_id, "Green-") end
+    "Mattapan Line" -> fn route_id -> route_id == "Mattapan" end
+    nil -> fn _ -> true end
+  end
+```
+
+### Filter Vehicle Positions
+
+```elixir
+# QUESTION: Can we assume that we'll always have a data point for a vehicle
+# being stopped_at each stop it serves?
+# Or do we need to do some inference:
+# - it was incoming_at before,
+# - now it's in_transit_to the next stop,
+# - therefore it must have stopped at this stop between data snapshots
```

Comment:
@fsaid90 do you know the answer to this definitively? My gut is we can rely on it being `stopped` and just based on the numbers it seems likely that's playing out, but I know you and Eric have gone deep in this data.

Reply:
I'm also leary of attempting to do deeper inference because I don't think a vehicle passing through a stop without stopping, (say on a non-revenue trip) should be considered a departure.

Reply:
Yes, we can assume the vehicle will be `stopped_at `each stop it serves.